File: solutions/haskell/raindrops/1/src/Raindrops.hs
module Raindrops (convert) where

factorNamePairs :: [(Int, String)]
factorNamePairs = [(3, "Pling"), (5, "Plang"), (7, "Plong")]

convert :: Int -> String
convert n =
  let get_factor_name factor num = if rem num (fst factor) == 0
                                   then snd factor
                                   else ""
      factor_fun_results = map get_factor_name
      factor_name_result = concatMap (\fun -> fun n)
                           $ factor_fun_results factorNamePairs
  in if not (null factor_name_result)
     then factor_name_result
     else show n

File: solutions/clojure/raindrops/1/src/raindrops.clj
(ns raindrops)

(def factor-pairs [{:factor 3 :name "Pling"}
                   {:factor 5 :name "Plang"}
                   {:factor 7 :name "Plong"}])

(defn factor-pair-funs
  "Takes 1 factor-name pair, returns a naming function."
  [factor-pair]
  (fn [possible-factor]
    (let [result (mod possible-factor (:factor factor-pair))]
      (if (zero? result) (:name factor-pair) ""))))

(defn convert
  "Converts a given number into a string based on its factors."
  [num]
  (let [tests (map factor-pair-funs factor-pairs)
        result (clojure.string/join (map (fn [fun] (fun num)) tests))]
    (if (empty? result) (str num) result)))

(comment
  (convert 28)
  (convert 30)
  (convert 34)
  )
